### ISSUE TITLE
URL-Based Queues

### ISSUE THREAD

**Issue by indolering** [leviwheatcroft/superagent-throttle#4]
I'd like to order my edits to records to ensure that the front-end remains consistent with the backend client but I don't want to block ALL requests.  Most REST API's use a `domain.tld/base/id` pattern, so we could just create a new queue for each URL.

**Comment by leviwheatcroft** [comment]
Hey thanks for this.. I'm not sure I really understand what you need though..

suppose a new queue was automatically created for each end  point, you'd still need a global rate limit, and you'd still need to fire off the requests in the order they were placed in the queues. So multiple queues alone wouldn't effect behaviour.

Would it be correct to say that you want a global rate limit, but a per endpoint concurrency limit ? so for endpoint 'a' you might set a concurrency limit of 1 request at a time, but for endpoints 'b' through 'd' each endpoint might have a higher concurrency limit.. but all queues would share the same rate of x requests per y ms.

Including this kind of functionality in the plugin, with an easy to configure / understand interface, would make the plugin way more complex than it is now, so I'm reluctant because I don't think it's a popular usecase. (but that's open for discussion? let me know what you think).

I think it would be easier (more declarative?) to create whichever queues you needed in your own code. But you'd have to split the api rate limit between however many queues you have, which might be impractical.

If there was only one endpoint requiring special treatment (endpoint 'a'), and you wanted to limit concurrency on that endpoint to 1, then you could simply run all requests through a single queue, but only enqueue a request to endpoint 'a' when the last request to endpoint 'a' has returned. That way, requests to endpoint 'a' would be one at a time, and would only be delayed where your traffic to endpoints 'b' through 'd' had hit the rate limits. I realise this means making a rudimentary queue in your own code, but for an unusual use case like this I think that might be the best option.

I haven't tested the following snippet, but it's just an example of how you might implement a very simple queue with a concurrency of one.

``` javascript
    const _ = require('superagent')
    const Throttle = require('superagent-throttle')
    let queue = []
    let enqueue
    let unqueue
    let throttle

    throttle = new Throttle({
      active: true,
      rate: 5,
      ratePer: 10000,
      concurrent: 2
    })

    // push something onto your queue
    enqueue = function(request, callback) {
      queue.push(function() {
        request.use(throttle.plugin) // your instance
        request.end(function() {
          unqueue() // next request if exists
          callback() // passed in callback
        })
      })
    }

    // fire next queued request
    unqueue = function() {
      if (queue.length == 0) {
        return // drained
      }
      queue.shift()() // fire oldest request in the queue
    }

    // how to queue up a request
    enqueue(
      request.get('http://foo.com/someUpdate/endpoint'),
      function(res) {
        // deal with response
      }
    )
```

So.. I'm doing my best to be helpful, but I realise I might have misunderstood your usecase. Let me know if this helps at all or maybe clarify exactly what you need :) Thanks again!

**Comment by indolering** [comment]
> suppose a new queue was automatically created for each end point, you'd still need a global rate limit, and you'd still need to fire off the requests in the order they were placed in the queues. So multiple queues alone wouldn't effect behaviour.

Right, I was thinking of a global queue with the existing rate limiting behavior but with subqueues that enforced strictly serial behavior.  I don't think the added complexity of per-sub queue behavior is worth the effort.

> I'm reluctant because I don't think it's a popular use-case. (but that's open for discussion? let me know what you think).

This is what every Ajax/client-side interface has to implement.  If you get edits to a specific record out-of-order, the client-side gets out of sync with the server side database.  The other issue you have to deal with is requests that don't make it through, but those trigger a refresh of the records.  So this would handle 90% of this specific use-case.

FWIW, there are jQuery plugins that allow developers to build queues for serial behavior.  They take the approach of specifying a name for each queue instead of building them based on the URL.

> I think it would be easier (more declarative?) to create whichever queues you needed in your own code. 

I've actually got a throttle function for creation and deletion of records, but the edit function is global and I thought it would be of better to add the functionality to the upstream library.

**Comment by leviwheatcroft** [comment]
ok so I've had a go at this, if you're interested have a look at branch `issue4` latest commit: 30f9a00f3b3950655957db315084752da8c10b81

this includes a breaking change, where the call before was `.use(throttle.plugin)` it now needs to be `.use(throttle.plugin())`

anyhow, you can create serialising subqueues on the fly with something like `.use(throttle.plugin('http://api.com/some/endpoint'))`, you don't need to pass a valid url, any string will do, it's just a namespace.

If you wanted to try out that branch that would be fantastic, otherwise I'll get around testing it sooner or later.

**Comment by leviwheatcroft** [comment]
I've just published v0.1.0 which adds functionality for serialised subthrottles. 

What I said above re the breaking change has been corrected, so the old style call still works, but I guess it's deprecated.

Thanks for the suggestion re serialising.
